app/serializers/api/cached_enterprise_serializer.rb

```diff
@@ -81,25 +81,42 @@ def supplied_properties
       end
     end
 
+    # This results in 3 queries per enterprise
     def distributed_properties
-      # This results in 3 queries per enterprise
+      (distributed_product_properties + distributed_producer_properties).uniq do |property_object|
+        property_object.property.presentation
+      end
+    end
 
+    def distributed_product_properties
       if active
-        product_properties = Spree::Property.currently_sold_by(object)
-        producer_property_ids = ProducerProperty.currently_sold_by(object).pluck(:property_id)
-
+        Spree::Property.currently_sold_by(enterprise)
       else
-        product_properties = Spree::Property.ever_sold_by(object)
-        producer_property_ids = ProducerProperty.ever_sold_by(object).pluck(:property_id)
+        Spree::Property
+          .joins(products: { variants: { exchanges: :order_cycle } })
+          .merge(Exchange.outgoing)
+          .merge(Exchange.to_enterprise(enterprise))
+          .select('DISTINCT spree_properties.*')
```

Comment:
I dont get this part. you just moved the code from Spree::Property.ever_sold_by to here, right?

Reply:
Yep, that's it. Essentially, `#distributed_properties` issues few queries to the properties table but this makes things more complex than they should be. It's simpler to just ask the DB about properties that we care for than having to deal with `ProducerProperty` on one hand, `product_properties` on the other, then mapping their presentation attribute, etc. We tend to forget that SQL is declarative and end up implementing map-reduce.

A first step towards relying on SQL is to inline the queries to find how they can be merged together. Does it make sense?

Reply:
yes it makes sense. but I wouldnt move/duplicate the SQL here if I was not changing it.